src/collectives.c

```diff
@@ -759,80 +759,75 @@ shmem_internal_op_to_all_recdbl_sw(void *target, const void *source, int count,
     /* extra peer exchange: grab information from extra_peer so its part of
      * pairwise exchange */
     if (my_id >= pow2_proc) {
-        peer = (my_id - pow2_proc) * stride + PE_start;
+        int peer = (my_id - pow2_proc) * stride + PE_start;
+
+        /* Wait for target ready, required when source and target overlap */
+        SHMEM_WAIT_UNTIL(pSync_extra_peer, SHMEM_CMP_EQ, ps_target_ready);
+
         shmem_internal_put_nb(target, current_target, wrk_size, peer,
                               &completion);
         shmem_internal_put_wait(&completion);
         shmem_internal_fence();
 
-        buff = neg_one;
-        shmem_internal_put_small(pSync_extra_peer, &buff, sizeof(long),
-                                 peer);
-        shmem_internal_fence();
-        buff = neg_one;
-        SHMEM_WAIT_UNTIL(pSync_extra_peer, SHMEM_CMP_EQ, buff);
+        shmem_internal_put_small(pSync_extra_peer, &ps_data_ready, sizeof(long), peer);
+        SHMEM_WAIT_UNTIL(pSync_extra_peer, SHMEM_CMP_EQ, ps_data_ready);
 
     } else {
-        if ((PE_size - pow2_proc) > my_id) {
-            peer = (my_id + pow2_proc) * stride + PE_start;
-            buff = neg_one;
-            SHMEM_WAIT_UNTIL(pSync_extra_peer, SHMEM_CMP_EQ, buff);
+        if (my_id < PE_size - pow2_proc) {
+            int peer = (my_id + pow2_proc) * stride + PE_start;
+            shmem_internal_put_small(pSync_extra_peer, &ps_target_ready, sizeof(long), peer);
 
+            SHMEM_WAIT_UNTIL(pSync_extra_peer, SHMEM_CMP_EQ, ps_data_ready);
             shmem_internal_reduce_local(op, datatype, count, target, current_target);
         }
 
         /* Pairwise exchange: (only for PE's that are within the power of 2
          * set) with every iteration, the information from each previous
          * exchange is passed forward in the new interation */
 
-        long *step_psync;
-
         for (i = 0; i < log2_proc; i++) {
+            long *step_psync = &pSync[i];
+            int peer = (my_id ^ (1 << i)) * stride + PE_start;
 
-            peer = (my_id ^ (1 << i)) * stride + PE_start;
-            step_psync = &pSync[i];
+            if (shmem_internal_my_pe < peer) {
+                shmem_internal_put_small(step_psync, &ps_target_ready,
```

Comment:
This isn't related to the source == target case.  We need this sync in every iteration to avoid overwriting the target buffer before the remote PE has consumed the data from the previous iteration (including in the first iteration to avoid overwriting data sent by one of the "extra" PEs).